euphoria-flink/src/main/java/cz/seznam/euphoria/flink/accumulators/FlinkNativeAccumulators.java

```diff
@@ -0,0 +1,175 @@
+/**
+ * Copyright 2016-2017 Seznam.cz, a.s.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package cz.seznam.euphoria.flink.accumulators;
+
+import cz.seznam.euphoria.core.client.accumulators.Accumulator;
+import cz.seznam.euphoria.core.client.accumulators.AccumulatorProvider;
+import cz.seznam.euphoria.core.client.accumulators.Counter;
+import cz.seznam.euphoria.core.client.accumulators.Histogram;
+import cz.seznam.euphoria.core.client.accumulators.Timer;
+import cz.seznam.euphoria.core.util.Settings;
+import org.apache.flink.api.common.accumulators.LongCounter;
+import org.apache.flink.api.common.functions.RuntimeContext;
+
+import java.time.Duration;
+import java.util.TreeMap;
+
+/**
+ * Integrates Flink native accumulator API.
+ */
+public class FlinkNativeAccumulators implements AccumulatorProvider {
+
+  private final RuntimeContext context;
+
+  private FlinkNativeAccumulators(RuntimeContext context) {
+    this.context = context;
+  }
+
+  @Override
+  public Counter getCounter(String name) {
+    System.out.println("Getting counter: " + name);
+    return getByNameAndClass(name, FlinkCounter.class);
+  }
+
+  @Override
+  public Histogram getHistogram(String name) {
+    return getByNameAndClass(name, FlinkHistogram.class);
+  }
+
+  @Override
+  public Timer getTimer(String name) {
+    return getByNameAndClass(name, FlinkTimer.class);
+  }
+
+  @SuppressWarnings("unchecked")
+  private <ACC extends Accumulator> ACC getByNameAndClass(String name,
+                                                                  Class<ACC> clz) {
+    Accumulator acc = (Accumulator) context.getAllAccumulators().get(name);
+
+    if (acc == null) {
+      try {
+        System.out.println("Creating counter: " + name);
+        // register a new instance
+        acc = clz.getConstructor().newInstance();
+        context.addAccumulator(name, (org.apache.flink.api.common.accumulators.Accumulator) acc);
+      } catch (Exception e) {
+        throw new RuntimeException("Exception during accumulator initialization: " + clz, e);
+      }
+    }
+
+    if (!clz.isAssignableFrom(acc.getClass())) {
```

Comment:
This looks expensive, and it will be called for each input element (if using counters). Can we move it into the `if (acc == null)` branch and perform the validation only once?